### PR TITLE
MOM6 IAU and atmos stochy restart test

### LINKED ISSUE #69: Add IAU capability to MOM6
Alex Bozec and Alan Wallcraft from COAPS/Florida State University have developed Incremental analysis update capability for MOM6. An older commit has been tested in the UFS-datm with a "replay" to ORAS5 analysis.  I will test it out the latest updates once hera come back online.

There will be an assoicate PR in the ufs-weather-model to modify the CMakeLists.txt file and add a IAU regression test to ocean model.

Draft PR is https://github.com/NOAA-EMC/MOM6/pull/68

### LINKED ISSUE #318: Restart runs with stochastic physics due not reproduce
## Description

Runs with stochastic physics turned on do not give b4b identical results in restart runs. This is because the initialization logic does not support restart runs, and the necessary fields aren't written to the restart fields (sppt/shum/skeb/... weights, ...)

### To Reproduce:
Use GFs v16beta regression test run directories to set up a 0->48h run, compare to a 0->24h + 24h->48h run. This does not reproduce. Turn off stochastic physics in all three runs to get b4b identical results (in PROD mode).

### LINKED ISSUE #667: Ocean model incremental analysis update (IAU)
## Description
 We need to implement IAU in the ocean in order to reduce initialization shocks during the data assimilation cycle
## Solution
 Alex Bozec at COAPS/FSU has coded up the solution in MOM6
## Alternatives
non needed.
## Related to
Directly reference any issues or PRs in this or other repositories that this is related to, and describe how they are related. Example:
- required to support MOM6 PR https://github.com/NOAA-EMC/MOM6/pull/68

### PR DESCRIPTION
**Pull request by pjpegion**
# PR Checklist

- [X] Ths PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR. Please consult the ufs-weather-model [wiki](https://github.com/ufs-community/ufs-weather-model/wiki/Making-code-changes-in-the-UFS-weather-model-and-its-subcomponents) if you are unsure how to do this.

- [X] This PR has been tested using a branch which is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR

- [X] An Issue describing the work contained in this PR has been created either in the subcomponent(s) or in the ufs-weather-model. The Issue should be created in the repository that is most relevant to the changes in contained in the PR. The Issue and the dependent sub-component PR 
are specified below.

- [ ] If new or updated input data is required by this PR, it is clearly stated in the text of the PR.

The newly added test will need a new input file 

@[INPUTDATA_ROOT]/MOM6_IC/@[OCNRES]/2011100100/mom6_increment.nc, this file currently resides on hera in /scratch2/BMC/gsienkf/Philip.Pegion/MOM6-IAU/mom6_increment.nc ./INPUT/

## Instructions: All subsequent sections of text should be filled in as appropriate.

The code changes are contained within MOM6.  One additional file was added to the MOM6 repo, thus the mom6_files.cmake
was updated.  

## Description

This PR brings in the ability for the ocean model to read in either analysis increments or a full analysis field and apply the increments over a specified time interval.  I new regression test was added for this.  In addition, a regression test was added to test the atmospheric stochastic physics restart capability.

### Issue(s) addressed

Link the issues to be closed with this PR, whether in this repository, or in another repository.
(Remember, issues must always be created before starting work on a PR branch!) 
- fixes https://github.com/ufs-community/ufs-weather-model/issues/667
- fixes https://github.com/NOAA-EMC/MOM6/issues/69
- 
## Testing

How were these changes tested? What compilers / HPCs was it tested with? Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) Have regression tests and unit tests (utests) been run? On which platforms and with which compilers? (Note that unit tests can only be run on tier-1 platforms)

- [x] hera.intel
- [x] hera.gnu
- [x] orion.intel
- [x] cheyenne.intel 
- [x] cheyenne.gnu
- [x] gaea.intel 
- [x] jet.intel
- [ ] wcoss_cray - not available
- [ ] wcoss_dell_p3 - not available
- [x] CI https://github.com/ufs-community/ufs-weather-model/actions/runs/1053575434

## Dependencies

If testing this branch requires non-default branches in other repositories, list them. Those branches should have matching names (ideally).

Do PRs in upstream repositories need to be merged first?
If so add the "waiting for other repos" label and list the upstream PRs
- waiting on https://github.com/NOAA-EMC/MOM6/pull/68

### CODE REVIEW

**Comment by DeniseWorthen** [comment]
@pjpegion I don't think a new ``MOM-IAU_input_template_100`` is required. You could implement the additional required parameters by having a new variable for ODA_INCUPD (set false by default). We did something similar when we removed the extra MOM template that included waves. You can look at how the variable ``MOM6_USE_WAVES`` works. 

When you want to include your feature, your test would the appropriate variable to true. I'd also suggest a variable for the ``ODA_INCUPD_NHOURS`` instead of a hard-wired "6". Since the value you're using is different than the default, I'm assuming it might be something which changes.

**Comment by pjpegion** [comment]
@DeniseWorthen I will make those changes. Thanks

**Comment by pjpegion** [comment]
I have removed the MOM-IAU_input_template_100 and added variables to MOM_input_template_100.

File: tests/fv3_conf/cpld_datm_gefs.IN [hunk 1]
```diff
@@ -22,4 +22,7 @@ if [[ $OCNRES == '025' ]]; then
 else
   cp    @[INPUTDATA_ROOT]/MOM6_IC/@[OCNRES]/2011100100/MOM6_IC_TS*.nc ./INPUT/MOM6_IC_TS.nc
   cp    @[INPUTDATA_ROOT]/CICE_IC/@[OCNRES]/cice_model_@[ICERES].cpc*.nc cice_model.res.nc
+  # MOM6 increment file
+  #cp    @[INPUTDATA_ROOT]/MOM6_IC/@[OCNRES]/2011100100/mom6_increment.nc ./INPUT/mom6_increment.nc
+  cp    /scratch2/BMC/gsienkf/Philip.Pegion/MOM6-IAU/mom6_increment.nc ./INPUT/mom6_increment.nc
```
Comment (DusanJovic-NOAA) [inline tests/fv3_conf/cpld_datm_gefs.IN]: Please revert this change and test on hera.

**Comment by DeniseWorthen** [comment]
I am going to remove the label for input data needed; the required file was added in the previous PR.

File: tests/fv3_conf/cpld_datm_gefs.IN (hunk 1 above)
Comment (pjpegion) [reply tests/fv3_conf/cpld_datm_gefs.IN]: testing now

File: tests/tests/datm_control_iau_gefs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#
+#  DATM_MOM6_CICE_CMEPS_GEFS test
+#
+
+export TEST_DESCR="DATM-MOM6-CICE-CMEPS_GEFS - IAU"
+
+export CNTL_DIR="datm_control_gefs"
```
Comment (DeniseWorthen) [inline tests/tests/datm_control_iau_gefs]: If you turn on MOM_IAU, do you expect this test to reproduce the datm_control_gefs?

Comment (pjpegion) [reply tests/tests/datm_control_iau_gefs]: No,  do I need to change CNTL_DIR to datm_control_iau_gefs?

Comment (DeniseWorthen) [reply tests/tests/datm_control_iau_gefs]: Yes, the CNTL_DIR is the test you are comparing against so change it to datm_control_iau_gefs.

**Comment by pjpegion** [comment]
There is a bugfix for stochastic physics that I am going to pull in with this PR.

File: tests/tests/control_stochy [hunk 3]
```diff
@@ -33,11 +33,12 @@ export IAER=5111
...
 export DO_SHUM=.T.
 export DO_SKEB=.T.
+export RESTART_INTERVAL=6
 export SKEB=0.3
 export SHUM=0.003
 export SPPT=0.2
 
 
-export FV3_RUN=control_run.IN
+export FV3_RUN=control_stochy_run.IN
```
Comment (DeniseWorthen) [inline tests/tests/control_stochy:42]: Does this run script exist? I don't see it in the current develop/tests/fv3_conf directory or in your PR branch.

Comment (pjpegion) [reply tests/tests/control_stochy:42]: hmm, maybe I didn't add it. Let me check.

**Comment by BrianCurtis-NOAA** [comment]
Machine: orion
Compiler: intel
Job: BL
Repo location: /work/noaa/nems/emc.nemspara/autort/pr/677884559/20210721120015/ufs-weather-model
Please manually delete: /work/noaa/stmp/bcurtis/stmp/bcurtis/FV3_RT/rt_114954
Test compile 007 failed failed 
Test compile 002 failed failed 
Test compile 009 failed failed 
Test compile 001 failed failed 
Test compile 003 failed failed 
Test compile 006 failed failed 
Test compile 005 failed failed 
Test compile 010 failed failed 
Test compile 004 failed failed 
Test compile 008 failed failed 
Test compile 012 failed failed 
Test compile 011 failed failed 
Test compile 017 failed failed 
Please make changes and add the following label back:
orion-intel-BL

**Comment by jiandewang** [comment]
wcoss is still down, I will launch RT once it's back

**Comment by DusanJovic-NOAA** [comment]
> Machine: orion
> Compiler: intel
> Job: BL
> Repo location: /work/noaa/nems/emc.nemspara/autort/pr/677884559/20210721120015/ufs-weather-model
> Please manually delete: /work/noaa/stmp/bcurtis/stmp/bcurtis/FV3_RT/rt_114954
> Test compile 007 failed failed
> Test compile 002 failed failed
> Test compile 009 failed failed
> Test compile 001 failed failed
> Test compile 003 failed failed
> Test compile 006 failed failed
> Test compile 005 failed failed
> Test compile 010 failed failed
> Test compile 004 failed failed
> Test compile 008 failed failed
> Test compile 012 failed failed
> Test compile 011 failed failed
> Test compile 017 failed failed
> Please make changes and add the following label back:
> orion-intel-BL

```
$ tail /work/noaa/stmp/bcurtis/stmp/bcurtis/FV3_RT/rt_114954/compile_001/err
CMake Error at CMakeLists.txt:144 (add_subdirectory):
  The source directory

    /work/noaa/nems/emc.nemspara/autort/pr/677884559/20210721120015/ufs-weather-model/stochastic_physics

  does not contain a CMakeLists.txt file.
```

**Comment by pjpegion** [comment]
I fixed the .gitmodules

**Comment by BrianCurtis-NOAA** [comment]
Machine: jet
Compiler: intel
Job: BL
Repo location: /lfs4/HFIP/h-nems/emc.nemspara/autort/pr/677884559/20210721170015/ufs-weather-model
Please manually delete: /lfs4/HFIP/h-nems/emc.nemspara/RT_RUNDIRS/emc.nemspara/FV3_RT/rt_285769
Baseline creation and move successful
Repo location: /lfs4/HFIP/h-nems/emc.nemspara/autort/pr/677884559/20210721185533/ufs-weather-model
Please manually delete: /lfs4/HFIP/h-nems/emc.nemspara/RT_RUNDIRS/emc.nemspara/FV3_RT/rt_280735
Test control_c384gdas 024 failed failed 
Test control_c384gdas 024 failed in run_test failed 
Please make changes and add the following label back:
jet-intel-BL

**Comment by DusanJovic-NOAA** [comment]
> Machine: jet
> Compiler: intel
> Job: BL
> Repo location: /lfs4/HFIP/h-nems/emc.nemspara/autort/pr/677884559/20210721170015/ufs-weather-model
> Please manually delete: /lfs4/HFIP/h-nems/emc.nemspara/RT_RUNDIRS/emc.nemspara/FV3_RT/rt_285769
> Baseline creation and move successful
> Repo location: /lfs4/HFIP/h-nems/emc.nemspara/autort/pr/677884559/20210721185533/ufs-weather-model
> Please manually delete: /lfs4/HFIP/h-nems/emc.nemspara/RT_RUNDIRS/emc.nemspara/FV3_RT/rt_280735
> Test control_c384gdas 024 failed failed
> Test control_c384gdas 024 failed in run_test failed
> Please make changes and add the following label back:
> jet-intel-BL

```
$ cat /lfs4/HFIP/h-nems/emc.nemspara/RT_RUNDIRS/emc.nemspara/FV3_RT/rt_280735/control_c384gdas/err
...
  0:  in fcst,init total time:    17.4488160610199     
181: [181:x331] unexpected DAPL connection event 0x4008 from 358
181: Fatal error in PMPI_Wait: Internal MPI error!, error stack:
181: PMPI_Wait(219)...........: MPI_Wait(request=0x1d694d20, status=0x1) failed
181: MPIR_Wait_impl(81).......: fail failed
181: PMPIDI_CH3I_Progress(850): fail failed
181: (unknown)(): Internal MPI error!
  0: slurmstepd: error: *** STEP 55251043.0 ON x9 CANCELLED AT 2021-07-21T20:02:04 ***
srun: Job step aborted: Waiting up to 32 seconds for job step to finish.
srun: error: x330: task 150: Killed
...
```

**Comment by BrianCurtis-NOAA** [comment]
Machine: jet
Compiler: intel
Job: RT
Repo location: /lfs4/HFIP/h-nems/emc.nemspara/autort/pr/677884559/20210721214508/ufs-weather-model
Please manually delete: /lfs4/HFIP/h-nems/emc.nemspara/RT_RUNDIRS/emc.nemspara/FV3_RT/rt_144572
Test control_2threads 018 failed failed 
Test control_2threads 018 failed in run_test failed 
Please make changes and add the following label back:
jet-intel-RT

**Comment by DusanJovic-NOAA** [comment]
> Machine: jet
> Compiler: intel
> Job: RT
> Repo location: /lfs4/HFIP/h-nems/emc.nemspara/autort/pr/677884559/20210721214508/ufs-weather-model
> Please manually delete: /lfs4/HFIP/h-nems/emc.nemspara/RT_RUNDIRS/emc.nemspara/FV3_RT/rt_144572
> Test control_2threads 018 failed failed
> Test control_2threads 018 failed in run_test failed
> Please make changes and add the following label back:
> jet-intel-RT

```
$ tail /lfs4/HFIP/h-nems/emc.nemspara/RT_RUNDIRS/emc.nemspara/FV3_RT/rt_144572/control_2threads/err
72:  Size of pset is:          104
72:  datset is: GFSFLX
72:  MXFLD is:          999
72:  size of lvlsxml:       150000
72:  size of post_avblflds param         104
72:  in readxml. nfld=         104 nrecout=         113
72:  post_fname=GFSFLX.GrbF24
srun: Job step aborted: Waiting up to 32 seconds for job step to finish.
 0: slurmstepd: error: *** STEP 55257334.0 ON x371 CANCELLED AT 2021-07-21T22:44:58 DUE TO TIME LIMIT ***
slurmstepd: error: *** JOB 55257334 ON x371 CANCELLED AT 2021-07-21T22:44:58 DUE TO TIME LIMIT ***
```

File: tests/tests/control_stochy (hunk 3 above)
Comment (junwang-noaa) [inline tests/tests/control_stochy:42]: It seems to me the only difference between the control_run.IN and control_stochy_run.IN is that the stochy is using 6hr restart files while control is using 12, maybe in the future we can add a restart file time to the script and unify the two scripts so that we don't need to maintain two scripts.

File: tests/fv3_conf/cpld_datm_gefs.IN [hunk 4]
```diff
@@ -22,4 +22,6 @@ if [[ $OCNRES == '025' ]]; then
 else
   cp    @[INPUTDATA_ROOT]/MOM6_IC/@[OCNRES]/2011100100/MOM6_IC_TS*.nc ./INPUT/MOM6_IC_TS.nc
   cp    @[INPUTDATA_ROOT]/CICE_IC/@[OCNRES]/cice_model_@[ICERES].cpc*.nc cice_model.res.nc
+  # MOM6 increment file
+  cp    @[INPUTDATA_ROOT]/MOM6_IC/@[OCNRES]/2011100100/mom6_increment.nc ./INPUT/mom6_increment.nc
```
Comment (junwang-noaa) [inline tests/fv3_conf/cpld_datm_gefs.IN:26]: I am curious, so the increments in this file will be applied to the corresponding MOM6 fields every MOM time step during the forecast time fh=0-MOM_IAU_HRS? How about restart runs, will it be applied from current starting time to current starting time +MOM_?IAU_HRS

**Review by DusanJovic-NOAA — APPROVED** [review]

**Comment by pjpegion** [comment]
submodules and gitmodules now point to the authoritative repos and is ready to merge.

**Review by DeniseWorthen — APPROVED** [review]

**Review by junwang-noaa — APPROVED** [review]

File: tests/tests/control_stochy (hunk 3 above)
Comment (pjpegion) [reply tests/tests/control_stochy:42]: Yes, that can be added so we don't have 2 scripts.

File: tests/fv3_conf/cpld_datm_gefs.IN (hunk 4 above)
Comment (pjpegion) [reply tests/fv3_conf/cpld_datm_gefs.IN:26]: Correct, that is the way I am doing the replay.
